src/planner/plan_window_aggregation.cpp

```diff
@@ -0,0 +1,118 @@
+#include <memory>
+#include <utility>
+
+#include "binder/bound_expression.h"
+#include "binder/bound_statement.h"
+#include "binder/bound_table_ref.h"
+#include "binder/expressions/bound_alias.h"
+#include "binder/expressions/bound_window.h"
+#include "binder/statement/select_statement.h"
+#include "binder/tokens.h"
+#include "common/exception.h"
+#include "common/macros.h"
+#include "common/util/string_util.h"
+#include "execution/expressions/abstract_expression.h"
+#include "execution/expressions/column_value_expression.h"
+#include "execution/expressions/constant_value_expression.h"
+#include "execution/plans/abstract_plan.h"
+#include "execution/plans/aggregation_plan.h"
+#include "execution/plans/filter_plan.h"
+#include "execution/plans/projection_plan.h"
+#include "fmt/format.h"
+#include "planner/planner.h"
+#include "type/type_id.h"
+#include "type/value_factory.h"
+
+namespace bustub {
+
+// TODO(chi): clang-tidy on macOS will suggest changing it to const reference. Looks like a bug.
+
+/* NOLINTNEXTLINE */
+auto Planner::PlanSelectWindow(const SelectStatement &statement, AbstractPlanNodeRef child) -> AbstractPlanNodeRef {
+  /* For window function we don't do two passes rewrites like planning normal aggregations.
+   *  Because standard sql does not allow using window function results in where clause, and
+   *  our implementation does not support having on window function. We assume window functions
+   *  only appear in select list, and we can plan them in one pass.
+   */
+  std::vector<AbstractExpressionRef> exprs;
+  std::vector<std::string> column_names;
+  ;
+  uint32_t window_agg_id = statement.select_list_.size();
+  const BoundWindow *window_call = nullptr;
+
+  for (uint32_t i = 0; i < statement.select_list_.size(); i++) {
+    const auto &item = statement.select_list_[i];
+    if (item->HasWindowFunction()) {
+      if (window_agg_id != statement.select_list_.size()) {
+        throw Exception("multiple window function is not supported yet");
```

Comment:
Currently it only supports one window-function in select list. But I can extend it to support multiple ones.

Reply:
Multiple window functions supported now